File: libraries/typescript/src/plugins/openai/handler.ts
// Copyright (c) 2026 Komaa DigiTech
// SPDX-License-Identifier: MIT

/**
 * The relay: one Microsoft Teams call on one side, one OpenAI Realtime session
 * on the other.
 *
 * Speech to speech, so there is no transcription step in the middle and the
 * model hears the caller's voice rather than a transcript of it. The one thing
 * this plugin must get right that the others do not is the rate: the
 * Realtime API speaks 24 kHz and the call speaks 16 kHz, and `agent.ts` owns
 * that conversion so nothing here has to think about it.
 *
 * The call capabilities come from `standin/callTools`, so a capability added to
 * the SDK reaches this plugin without an edit. Add your own with
 * {@link CustomTool} and they are declared the same way.
 */

import type { CallHandler, CallSession } from "../../handler.js";
import { ExpressionCue } from "../../avatar.js";
import { CallTools } from "../../callTools.js";
import { fetchPublicImage } from "../../fetch.js";
import { TurnLipSync } from "../../lipsync.js";
import { logger } from "../../log.js";
import { StartupBuffer } from "../../startup.js";
import { FrameDescriber, type VideoFrame } from "../../vision.js";
import { VisionTools } from "../../visionTools.js";
import {
  AgentSocket,
  BUILT_IN_TOOL_NAMES,
  buildInstructions,
  buildSessionUpdate,
  type AgentMessage,
  type AgentPort,
  type CallerContext,
} from "./agent.js";
import { openAIConfigFromEnv, type OpenAIConfig } from "./config.js";

const MAX_EMOTION_CHARS = 40;
const MAX_CAPTION_CHARS = 200;
const IMAGE_FETCH_TIMEOUT_MS = 10_000;
const MAX_PENDING_AUDIO = 200;
const MAX_PENDING_CONTEXT = 20;

/** What a custom tool is told about the call it is running inside. */
export interface ToolContext {
  call: CallSession;
  /** Whether the call is being recorded. Gate anything that stores what the caller showed. */
  recording: boolean;
}

/** A function of your own, which the model calls and your code answers. */
export interface CustomTool {
  name: string;
  description: string;
  parameters?: Record<string, unknown>;
  handler: (
    params: Record<string, unknown>,
    ctx: ToolContext,
  ) => string | Promise<string>;
}

/** How the handler reaches OpenAI. Replaceable so tests need no network. */
export type Connect = (
  config: OpenAIConfig,
  onMessage: (message: AgentMessage) => void | Promise<void>,
  onAudio: (pcm: Buffer) => void | Promise<void>,
  onClose: (code: number, reason: string) => void | Promise<void>,
) => Promise<AgentPort>;

/** Options for {@link OpenAIHandler}. */
export interface OpenAIHandlerOptions {
  config?: OpenAIConfig;
  tools?: CustomTool[];
  /** Remote MCP entries, normalised with `mcpTool`. Run server-side by OpenAI. */
  mcpTools?: Array<Record<string, unknown>>;
  describer?: FrameDescriber;
  connect?: Connect;
}

/** One Microsoft Teams call answered by one OpenAI Realtime session. */
export class OpenAIHandler implements CallHandler {
  #config: OpenAIConfig;
  #tools: Map<string, CustomTool>;
  #mcpTools: Array<Record<string, unknown>>;
  #describer: FrameDescriber | undefined;
  #connect: Connect;
  #call: CallSession | undefined;
  #vision: VisionTools | undefined;
  #callTools: CallTools | undefined;
  #agent: AgentPort | undefined;
  #closed = false;
  /** The mouth and the face. Both cosmetic, both cheap, both per call. */
  readonly #lip = new TurnLipSync();
  readonly #cue = new ExpressionCue();
  #replyText = "";
  // Holds BOTH the caller's first words and the first context, so neither
  // is lost while the provider is still connecting.
  readonly #pending = new StartupBuffer();

  constructor(options: OpenAIHandlerOptions = {}) {
    this.#config = options.config ?? openAIConfigFromEnv();
    this.#tools = new Map(
      (options.tools ?? []).map((tool) => [tool.name, tool]),
    );
    this.#mcpTools = options.mcpTools ?? [];
    this.#describer = options.describer ?? FrameDescriber.fromEnv();
    this.#connect = options.connect ?? AgentSocket.connect.bind(AgentSocket);

    const collisions = [...this.#tools.keys()].filter((name) =>
      BUILT_IN_TOOL_NAMES.has(name),
    );
    if (collisions.length > 0) {
      throw new Error(
        `custom tools may not shadow built-in ones: ${collisions.sort().join(", ")}`,
      );
    }
  }

  /**
   * Whether the call is being recorded, straight off the session.
   *
   * The server keeps this current from `session.start` and every later
   * `recording.status`, so there is nothing to re-derive here.
   */
  #recording(): boolean {
    return this.#call?.recordingActive ?? false;
  }

  // ---- the SDK seam -------------------------------------------------------

  async onStart(session: CallSession): Promise<void> {
    this.#call = session;
    this.#vision = new VisionTools(session, { describer: this.#describer });
    this.#callTools = new CallTools(session, { vision: this.#vision });
    const caller = session.start.caller;
    const context: CallerContext = {
      callerName: caller.displayName ?? "the caller",
      tenantId: caller.tenantId ?? "unknown-tenant",
      direction: session.start.direction,
    };

    let agent: AgentPort;
    try {
      agent = await this.#connect(
        this.#config,
        (message) => this.#onAgentMessage(message),
        (pcm) => this.#onAgentAudio(pcm),
        (code, reason) => this.#onAgentClose(code, reason),
      );
    } catch (err) {
      logger.error(
        `standin: could not open the OpenAI Realtime session: ${String(err)}`,
      );
      await session.end("agent-unavailable");
      return;
    }

    // The call can end during the connect above.
    if (this.#closed) {
      await agent.aclose();
      return;
    }
    this.#agent = agent;

    const extraTools = [
      ...[...this.#tools.values()].map((tool) => ({
        type: "function",
        name: tool.name,
        description: tool.description,
        parameters: tool.parameters ?? {
          type: "object",
          properties: {},
          required: [],
        },
      })),
      ...this.#mcpTools,
    ];
    agent.sendSessionUpdate(
      buildSessionUpdate(
        this.#config,
        buildInstructions(this.#config.instructions, context),
        extraTools,
      ),
    );
    await this.#pending.release(
      (pcm: Buffer) => agent.sendAudio(pcm),
      (text: string) => agent.sendContext(text),
    );
    if (this.#pending.dropped.audio || this.#pending.dropped.context) {
      logger.info(
        "standin: the caller outran the agent starting up; some early input was dropped",
      );
    }
  }

  async onCallerAudio(pcm: Buffer): Promise<void> {
    const agent = this.#agent;
    if (agent === undefined || !agent.isOpen) {
      this.#pending.audio(pcm);
      return;
    }
    agent.sendAudio(pcm);
  }

  /**
   * Context arrives as a user item with no response requested, so the model
   * reads it on its next turn instead of interrupting to acknowledge it.
   */
  /**
   * Keep a short history, so a question about a slide already gone can still be
   * answered. Only kept while the call is recorded.
   */
  async onVideoFrame(frame: VideoFrame): Promise<void> {
    this.#vision?.keyframes.offer(frame, this.#call?.recordingActive ?? false);
  }

  async onContext(text: string): Promise<void> {
    const agent = this.#agent;
    if (agent === undefined || !agent.isOpen) {
      this.#pending.context(text);
      return;
    }
    agent.sendContext(text);
  }

  /** Cancel whatever is being said, then ask for the goodbye. */
  async onGoodbye(text: string): Promise<void> {
    const agent = this.#agent;
    if (agent === undefined || !agent.isOpen) return;
    agent.cancelResponse();
    agent.sendUserTurn(
      `[system: the call is ending. Say a brief goodbye now: "${text}"]`,
    );
  }

  async aclose(): Promise<void> {
    this.#closed = true;
    const agent = this.#agent;
    this.#agent = undefined;
    if (agent !== undefined) {
      try {
        await agent.aclose();
      } catch {
        // Teardown must not throw: the slot is freed either way.
      }
    }
  }

  /**
   * The shared vision tools for this call.
   *
   * They carry the budget, the keyframes and the graceful refusals, so this
   * plugin no longer keeps its own copy of any of them.
   */
  #visionTools(): VisionTools {
    if (this.#vision === undefined)
      throw new Error("the call has not started yet");
    return this.#vision;
  }

  // ---- what OpenAI sends us -----------------------------------------------

  /** Agent audio, already resampled to the wire rate by the socket. */
  async #onAgentAudio(pcm: Buffer): Promise<void> {
    if (pcm.length === 0) return;
    await this.#call?.sendAudio(pcm);
    // The only duration this worker genuinely knows. A realtime model hands
    // back no timings, and a guess from text length drifts further out of step
    // with the voice the longer the turn runs.
    this.#lip.audioSent(pcm);
  }

  /**
   * One viseme timeline per turn, spread over the audio that turn actually sent.
   *
   * Cosmetic, so everything here is swallowed: the worst acceptable outcome is
   * a still mouth over correct audio, and the unacceptable one is a dropped
   * turn because a lip shape threw.
   */
  async #sendVisemes(): Promise<void> {
    const text = this.#replyText;
    this.#replyText = "";
    try {
      const marks = this.#lip.finish(text);
      if (marks.length > 0) await this.#call?.sendSpeechMarks(marks);
    } catch (err) {
      logger.debug(`standin: the viseme timeline was skipped: ${String(err)}`);
    }
  }

  /**
   * A piece of what the model is saying, as it is being said.
   *
   * Held for the viseme timeline, and read for the face. Re-read on every piece
   * rather than at the end, because waiting for the final transcript leaves the
   * face wrong for the whole time the reply is being spoken.
   */
  async #onReplyText(text: string): Promise<void> {
    if (text === "") return;
    this.#replyText += text;
    try {
      const emotion = this.#cue.cue(this.#replyText);
      if (emotion !== null) await this.#call?.express(emotion);
    } catch (err) {
      logger.debug(`standin: the expression cue was skipped: ${String(err)}`);
    }
  }

  async #onAgentMessage(message: AgentMessage): Promise<void> {
    switch (message.type) {
      case "input_audio_buffer.speech_started":
        // The model has already cancelled its own response server-side. This is
        // the other half: flushing what StandIn has buffered, which is what
        // actually stops the bot mid-word.
        await this.#call?.cancelPlayback();
        // The service drops audio the caller never heard. Keeping the count
        // would spread the NEXT turn's words over its own audio plus the audio
        // that was thrown away, and the mouth would run long for the rest of it.
        this.#lip.cancel();
        this.#replyText = "";
        break;
      case "response.output_audio_transcript.delta":
        await this.#onReplyText(
          typeof message.delta === "string" ? message.delta : "",
        );
        break;
      case "response.done":
        await this.#sendVisemes();
        break;
      case "response.function_call_arguments.done":
        void this.#runTool(message);
        break;
      case "conversation.item.input_audio_transcription.completed":
      case "response.output_audio_transcript.done":
        if (this.#config.logTranscripts && this.#recording()) {
          logger.info(`standin: openai ${message.type}`);
        }
        break;
      case "error":
        logger.warn(
          `standin: OpenAI error: ${JSON.stringify(message.error ?? message)}`,
        );
        break;
      default:
        break;
    }
  }

  async #onAgentClose(code: number, reason: string): Promise<void> {
    logger.info(
      `standin: the OpenAI Realtime session closed (${code} ${reason})`,
    );
    if (!this.#closed) await this.#call?.end("agent-disconnected");
  }

  // ---- the model's tools --------------------------------------------------

  async #runTool(message: AgentMessage): Promise<void> {
    const name = message.name;
    const callId = message.call_id;
    if (typeof name !== "string" || typeof callId !== "string") return;

    let params: Record<string, unknown> = {};
    if (typeof message.arguments === "string") {
      try {
        const parsed: unknown = JSON.parse(message.arguments || "{}");
        if (typeof parsed === "object" && parsed !== null) {
          params = parsed as Record<string, unknown>;
        }
      } catch {
        params = {};
      }
    }

    let output: string;
    try {
      output = await this.dispatch(name, params);
    } catch (err) {
      // Tell the model the truth: a tool that silently "succeeded" leaves it
      // talking about something that never happened.
      output = `${name} failed: ${String(err)}`;
    }
    this.#agent?.sendToolResult(callId, output);
  }

  /** Run one tool and return what the model is told. Exposed for tests. */
  async dispatch(
    name: string,
    params: Record<string, unknown>,
  ): Promise<string> {
    const call = this.#call;
    if (call === undefined) return "the call is no longer active";

    const callTools = this.#callTools;
    if (callTools !== undefined && BUILT_IN_TOOL_NAMES.has(name)) {
      // The SDK owns these, and its dispatch never throws: what comes back is
      // the sentence the model should say.
      return callTools.dispatch(name, params);
    }

    const tool = this.#tools.get(name);
    if (tool === undefined)
      return `"${name}" is not a tool this plugin answers`;
    return String(
      await tool.handler(params, { call, recording: this.#recording() }),
    );
  }
}

File: libraries/python/standin/plugins/elevenlabs/handler.py
# Copyright (c) 2026 Komaa DigiTech
# SPDX-License-Identifier: MIT

"""The relay: one Microsoft Teams call on one side, one ElevenLabs agent on the other.

This is a plain :class:`~standin.CallHandler`. Everything that is the same for
every framework - the socket StandIn dials, the handshake, capacity, the frame
loop, the watchdogs - belongs to :class:`~standin.CallServer`, so what is left
here is only what ElevenLabs specifically needs:

* open a conversation when the call starts, personalised with the caller
* forward the caller's voice, and play the agent's voice back
* stop talking the instant ElevenLabs reports the caller interrupted
* answer the agent's client tools with what a Microsoft Teams call can actually do

The client tools it answers are the SDK's call capabilities, so this agent can
do what every other provider's can. ElevenLabs declares client tools on the
agent rather than over the wire, so :func:`client_tools` prints the exact
declarations to paste in rather than leaving you to retype them.

Two of them mean something slightly different here, and the difference is the
reason they are not the SDK's implementations. ``look`` uploads the frame into
the conversation for the model to read directly, which PERSISTS the caller's
screen with ElevenLabs and is why it takes a recorded call. ``show_image`` also
accepts inline bytes, which no other provider offers.
"""

from __future__ import annotations

import asyncio
import base64
import contextlib
from typing import Any

from standin import CallSession
from standin.calltools import tool_schemas
from standin.fetch import fetch_public_image
from standin.log import logger
from standin.startup import StartupBuffer
from standin.vision import DISPLAY_IMAGE_MIME_TYPES, MAX_IMAGE_BYTES, VideoFrame
from standin.vision_tools import KeyframeStore

from .agent import AgentSocket, build_conversation_init
from .config import ElevenLabsConfig

__all__ = ["ElevenLabsHandler", "client_tools"]

#: Bound on what a model may put in an emotion or a caption. These reach the
#: avatar tile, and an unbounded string from a model is an unbounded string
#: from whoever is steering it.
_MAX_EMOTION_CHARS = 40
_MAX_CAPTION_CHARS = 200
_MAX_MODE_CHARS = 20

#: How long to wait for an agent-supplied URL to produce an image.
_IMAGE_FETCH_TIMEOUT_MS = 10_000


def client_tools() -> list[dict[str, Any]]:
    """The client tools to declare on the ElevenLabs agent.

    ElevenLabs configures tools on the agent, not over the wire, so there is
    nothing this plugin can send. What it can do is hand you the declarations
    verbatim::

        python -c "import json, standin.plugins.elevenlabs as e; \
            print(json.dumps(e.client_tools(), indent=2))"

    The names and descriptions are the SDK's, so this agent is told about the
    same capabilities as every other provider's. ``show_image`` is widened here
    because ElevenLabs is the one provider that can take the bytes inline.
    """
    tools = [dict(spec) for spec in tool_schemas("flat")]
    for tool in tools:
        if tool["name"] != "show_image":
            continue
        tool["description"] = (
            f"{tool['description']} You can give the image inline instead, "
            "as 'dataBase64' with its 'mime'."
        )
        parameters = {**tool["parameters"], "properties": {**tool["parameters"]["properties"]}}
        parameters["properties"].update(
            {
                "dataBase64": {
                    "type": "string",
                    "description": "The image itself, base64 encoded. An alternative to 'url'.",
                },
                "mime": {
                    "type": "string",
                    "description": f"Required with dataBase64: one of "
                    f"{', '.join(DISPLAY_IMAGE_MIME_TYPES)}.",
                },
                "durationMs": {
                    "type": "integer",
                    "description": "How long to leave it up, in milliseconds.",
                },
            }
        )
        # Either form will do, so neither argument can be required.
        parameters["required"] = []
        tool["parameters"] = parameters
    return tools


#: Caller audio and context that arrive while the conversation is still opening.
#: Bounded: on a socket that never opens these would otherwise grow for the
#: length of the call.
_MAX_PENDING_AUDIO = 200
_MAX_PENDING_CONTEXT = 20


class ElevenLabsHandler:
    """One Microsoft Teams call answered by one ElevenLabs agent."""

    def __init__(self, config: ElevenLabsConfig | None = None) -> None:
        self._config = config or ElevenLabsConfig.from_env()
        self._call: CallSession | None = None
        self._agent: AgentSocket | None = None
        self._closed = False
        # Holds BOTH the caller's first words and the first context, so
        # neither is lost while the provider is still connecting.
        self._pending = StartupBuffer()
        # Audio already in flight when an interruption lands is audio the caller
        # must never hear: the model stopped, and playing the tail would talk
        # over the person who just interrupted.
        self._last_audio_event = 0
        self._last_interrupt_event = 0
        self._tasks: set[asyncio.Task[Any]] = set()
        # Frames kept so a question about a slide already gone can still be
        # answered. The store keeps nothing unless the call is recorded.
        self._keyframes = KeyframeStore()

    def _recording(self) -> bool:
        """Whether the call is being recorded, straight off the session.

        The server keeps this current from ``session.start`` and every later
        ``recording.status``, so there is nothing to re-derive here.
        """
        return self._call is not None and self._call.recording_active

    # ---- the SDK seam -----------------------------------------------------

    async def on_start(self, session: CallSession) -> None:
        """Open the conversation, personalised with who is calling."""
        self._call = session
        caller = session.start.caller

        try:
            agent = await AgentSocket.connect(
                self._config, self._on_agent_message, self._on_agent_close
            )
        except Exception as err:
            logger.error("standin: could not open the ElevenLabs conversation: %s", err)
            await session.end("agent-unavailable")
            return

        # The call can end DURING the connect above. Keeping a socket opened
        # after teardown would leave a live, billed conversation with nothing on
        # the other end of it.
        if self._closed:
            await agent.aclose()
            return
        self._agent = agent

        agent.send_conversation_init(
            build_conversation_init(
                dynamic_variables={
                    "caller_name": caller.display_name or "caller",
                    "tenant_id": caller.tenant_id or "unknown-tenant",
                    "call_direction": session.start.direction,
                },
                first_message=self._config.first_message,
                environment=self._config.environment,
                # Per-person memory, and only when the person is actually
                # identified. Guests and anonymous callers get no user_id at
                # all, so two of them can never share one identity.
                user_id=caller.aad_id,
                branch_id=self._config.agent_branch_id,
            )
        )

        # Whatever arrived while the socket was opening. The "there are N people
        # here, stay quiet" signal usually lands exactly in this window.
        released = await self._pending.release(
            send_audio=lambda pcm: agent.send_audio_chunk(base64.b64encode(pcm).decode("ascii")),
            send_context=agent.send_contextual_update,
        )
        if any(self._pending.dropped):
            logger.info(
                "standin: the caller outran the agent starting up; some early input was dropped"
            )
        del released

    async def on_caller_audio(self, pcm: bytes) -> None:
        """The caller's voice, straight through. Both sides are PCM16 at 16 kHz."""
        chunk = base64.b64encode(pcm).decode("ascii")
        agent = self._agent
        if agent is None or not agent.is_open:
            self._pending.audio(pcm)
            return
        agent.send_audio_chunk(chunk)

    async def on_video_frame(self, frame: VideoFrame) -> None:
        """Keep a short history, so ``look_back`` has something to look at.

        The store itself refuses to keep anything unless the call is being
        recorded, which is the same promise ``look`` makes.
        """
        call = self._call
        if call is not None:
            self._keyframes.offer(frame, call.recording_active)

    async def on_context(self, text: str) -> None:
        """Participant counts, key presses, recording changes.

        Sent as a contextual update, which ElevenLabs delivers WITHOUT
        interrupting the agent mid-sentence.
        """
        agent = self._agent
        if agent is None or not agent.is_open:
            self._pending.context(text)
            return
        agent.send_contextual_update(text)

    async def on_goodbye(self, text: str) -> None:
        """StandIn is ending the call and wants this line spoken first.

        Delivered as a user turn, which interrupts whatever the agent was
        saying. The SDK has already flushed the buffered playback, so the line
        is not queued behind the answer it interrupted.
        """
        agent = self._agent
        if agent is None or not agent.is_open:
            return
        self._last_interrupt_event = max(self._last_interrupt_event, self._last_audio_event)
        agent.send_user_message(f'[system: the call is ending. Say a brief goodbye now: "{text}"]')

    async def aclose(self, reason: str) -> None:
        """Close the conversation. Always runs exactly once."""
        self._closed = True
        for task in list(self._tasks):
            task.cancel()
        agent, self._agent = self._agent, None
        if agent is not None:
            with contextlib.suppress(Exception):
                await agent.aclose()

    # ---- what ElevenLabs sends us -----------------------------------------

    async def _on_agent_message(self, message: dict[str, Any]) -> None:
        kind = message["type"]
        if kind == "audio":
            await self._on_agent_audio(message)
        elif kind == "interruption":
            await self._on_interruption(message)
        elif kind == "ping":
            event = message.get("ping_event")
            if isinstance(event, dict) and isinstance(event.get("event_id"), (int, float)):
                agent = self._agent
                if agent is not None:
                    agent.send_pong(int(event["event_id"]))
        elif kind == "client_tool_call":
            call = message.get("client_tool_call")
            if (
                isinstance(call, dict)
                and isinstance(call.get("tool_name"), str)
                and isinstance(call.get("tool_call_id"), str)
            ):
                await self._on_tool_call(call)
        elif kind in ("user_transcript", "agent_response"):
            # Gated twice. A transcript in your logs is a recording of the
            # caller, so it takes both an explicit opt-in AND the call actually
            # being recorded.
            if self._config.log_transcripts and self._recording():
                logger.info("standin: elevenlabs %s", kind)

    async def _on_agent_audio(self, message: dict[str, Any]) -> None:
        event = message.get("audio_event")
        if (
            not isinstance(event, dict)
            or not isinstance(event.get("event_id"), (int, float))
            or not isinstance(event.get("audio_base_64"), str)
        ):
            return
        event_id = int(event["event_id"])
        self._last_audio_event = max(self._last_audio_event, event_id)
        if event_id <= self._last_interrupt_event:
            # Audio the model generated before it was interrupted, arriving
            # after. Playing it is exactly the thing barge-in exists to stop.
            return
        call = self._call
        if call is None:
            return
        try:
            pcm = base64.b64decode(event["audio_base_64"], validate=True)
        except Exception:
            logger.warning("standin: ElevenLabs sent unusable audio; dropping the frame")
            return
        await call.send_audio(pcm)

    async def _on_interruption(self, message: dict[str, Any]) -> None:
        event = message.get("interruption_event")
        if not isinstance(event, dict) or not isinstance(event.get("event_id"), (int, float)):
            return
        self._last_interrupt_event = max(self._last_interrupt_event, int(event["event_id"]))
        call = self._call
        if call is not None:
            # The only lever that un-sends audio StandIn already has buffered.
            # Without it the model stops but the bot keeps talking.
            await call.cancel_playback()

    async def _on_agent_close(self, code: int, reason: str) -> None:
        logger.info("standin: ElevenLabs conversation closed (%s %s)", code, reason)
        call = self._call
        if call is not None and not self._closed:
            await call.end("agent-disconnected")

    # ---- the agent's client tools -----------------------------------------

    async def _on_tool_call(self, call: dict[str, Any]) -> None:
        params = call.get("parameters")
        if not isinstance(params, dict):
            params = {}
        name = call["tool_name"]
        tool_call_id = call["tool_call_id"]

        if name == "end_call":
            self._reply(tool_call_id, "the call is ending")
            if self._call is not None:
                await self._call.end("agent-ended-call")
        elif name == "express":
            await self._on_express(tool_call_id, params)
        elif name == "show_image":
            self._spawn(self._on_show_image(tool_call_id, params))
        elif name == "look":
            self._spawn(self._on_look(tool_call_id, params))
        elif name == "look_back":
            self._spawn(self._on_look_back(tool_call_id, params))
        else:
            self._reply(tool_call_id, f'"{name}" is not a tool this plugin answers', True)

    async def _on_express(self, tool_call_id: str, params: dict[str, Any]) -> None:
        emotion = params.get("emotion")
        emotion = emotion.strip() if isinstance(emotion, str) else ""
        if not emotion:
            self._reply(tool_call_id, "express needs an 'emotion'", True)
            return
        if self._call is not None:
            try:
                await self._call.express(emotion)
            except ValueError as err:
                # The bound lives where the message is built, so every plugin
                # gets it. Read it back rather than raising at the model.
                self._reply(tool_call_id, str(err), True)
                return
        self._reply(tool_call_id, f"expressing {emotion}")

    async def _on_show_image(self, tool_call_id: str, params: dict[str, Any]) -> None:
        """Put a picture on the bot's tile, from inline bytes or a URL."""
        call = self._call
        if call is None:
            return
        try:
            data_base64 = params.get("dataBase64")
            data_base64 = data_base64 if isinstance(data_base64, str) and data_base64 else None
            mime = params.get("mime") if isinstance(params.get("mime"), str) else None
            url = params.get("url") if isinstance(params.get("url"), str) else None

            if data_base64 is None and url:
                # The URL came from the model, which is steered by whoever is on
                # the call, so it is untrusted input. fetch_public_image refuses
                # private and link-local hosts and re-checks the address at
                # connect time, which is what stops a crafted prompt reaching
                # cloud metadata.
                image, mime = await fetch_public_image(
                    url, MAX_IMAGE_BYTES, _IMAGE_FETCH_TIMEOUT_MS
                )
                data_base64 = base64.b64encode(image).decode("ascii")

            if not data_base64 or mime not in DISPLAY_IMAGE_MIME_TYPES:
                raise ValueError(
                    "show_image needs {dataBase64, mime} or {url}, "
                    f"and the image must be one of {', '.join(DISPLAY_IMAGE_MIME_TYPES)}"
                )

            duration = params.get("durationMs")
            mode = params.get("mode")
            caption = params.get("caption")
            await call.display_image(
                data_base64,
                mime,
                duration_ms=int(duration) if isinstance(duration, (int, float)) else None,
                mode=mode[:_MAX_MODE_CHARS] if isinstance(mode, str) else None,
                caption=caption[:_MAX_CAPTION_CHARS] if isinstance(caption, str) else None,
            )
            self._reply(tool_call_id, "the caller can see the image")
        except Exception as err:
            # Tell the agent the truth. Claiming success would leave it talking
            # about a picture the caller never saw.
            self._reply(tool_call_id, f"show_image failed: {err}", True)

    async def _on_look(self, tool_call_id: str, params: dict[str, Any]) -> None:
        """Look at what the caller is showing.

        The frame is uploaded to the conversation, which PERSISTS the caller's
        screen or face with ElevenLabs. That is why it takes the call being
        recorded: the caller has been told the call is being kept, and this is
        part of what is kept.
        """
        call, agent = self._call, self._agent
        if call is None or agent is None:
            return
        source = params.get("source")
        frame = call.latest_video_frame(source if source in ("camera", "screenshare") else None)
        if frame is None:
            self._reply(
                tool_call_id,
                "there is nothing to look at: the caller is not sharing their camera or screen",
                True,
            )
            return
        if not self._recording():
            self._reply(
                tool_call_id,
                "cannot look: the Microsoft Teams call is not being recorded, and looking would "
                "store the caller's screen with a third party",
                True,
            )
            return

        await self._attach(tool_call_id, frame, params, "look")

    async def _on_look_back(self, tool_call_id: str, params: dict[str, Any]) -> None:
        """Look at something the caller has already moved past.

        Only possible on a recorded call, because that is the only time frames
        are kept at all.
        """
        if self._call is None or self._agent is None:
            return
        frames = self._keyframes.recent()
        if not frames:
            self._reply(
                tool_call_id,
                "I can only look back at earlier screens while the call is being recorded, "
                "and nothing has been kept"
                if not self._recording()
                else "nothing has been shown on this call yet",
                True,
            )
            return
        await self._attach(tool_call_id, frames[-1], params, "look_back")

    async def _attach(
        self, tool_call_id: str, frame: VideoFrame, params: dict[str, Any], tool: str
    ) -> None:
        """Upload one frame into the conversation and tell the agent to read it."""
        agent = self._agent
        if agent is None:
            return
        question = params.get("question")
        question = question.strip() if isinstance(question, str) else ""
        question = question or "Describe what is visible."
        who = frame.participant_name or (
            "a participant" if frame.source == "screenshare" else "the caller"
        )
        described = (
            f"screen shared by {who}" if frame.source == "screenshare" else f"camera of {who}"
        )
        try:
            await agent.attach_image(
                frame.data, frame.mime, f"[live call frame: {described}] {question}"
            )
            self._reply(tool_call_id, "the frame is attached; answer from what you can see")
        except Exception as err:
            self._reply(tool_call_id, f"{tool} failed: {err}", True)

    # ---- plumbing ---------------------------------------------------------

    def _reply(self, tool_call_id: str, result: str, is_error: bool = False) -> None:
        agent = self._agent
        if agent is not None:
            agent.send_tool_result(tool_call_id, result, is_error)

    def _spawn(self, coro: Any) -> None:
        """Run a tool off the read loop, keeping a reference so it survives.

        A tool that fetches a URL or uploads a frame takes seconds. Awaiting it
        inline would stall every other message from the agent, including the
        interruption that stops it talking.
        """
        task = asyncio.ensure_future(coro)
        self._tasks.add(task)
        task.add_done_callback(self._tasks.discard)
